### PR TITLE
Bau ser 02 upload azure

### PR DESCRIPTION
**Pull request by kduy**
Leave claim and claim photo untested but working

### CODE REVIEW

File: app/controllers/api/v1/claim_controller.rb [hunk 1]
```diff
@@ -0,0 +1,55 @@
+class Api::V1::ClaimController < ApplicationController
+  before_action :set_claim, except: [:create]
```
Comment (vsrboth) [inline app/controllers/api/v1/claim_controller.rb]: there should be another empty new line below this codes.
example,
```ruby
before_action :set_claim, except: [:create]

def show
```

File: app/controllers/api/v1/claim_controller.rb [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+  def create
+    @claim = Claim.new(claim_params)
+
+    @claim.save
+    render json: @claim,
+           serializer: ClaimSerializer,
+           root: 'claim'
+  end
+
+  def edit
```
Comment (vsrboth) [inline app/controllers/api/v1/claim_controller.rb]: prefer `def edit; end`

File: app/controllers/api/v1/claim_photo_controller.rb [hunk 3]
```diff
@@ -0,0 +1,42 @@
+class Api::V1::ClaimPhotoController < ApplicationController
+  before_action :set_claim_photo, except: [:create]
```
Comment (vsrboth) [inline app/controllers/api/v1/claim_photo_controller.rb]: there should be an api where I can fetch all the photo for this claim. what do u think?

File: app/controllers/api/v1/claim_controller.rb [hunk 4]
```diff
@@ -0,0 +1,55 @@
+class Api::V1::ClaimController < ApplicationController
```
Comment (vsrboth) [inline app/controllers/api/v1/claim_controller.rb]: prefer the following for declaring class to spot immediately that the class is under which folder. example,

```ruby
module Api
end
```

Comment (vsrboth) [inline app/controllers/api/v1/claim_controller.rb]: prefer plural for all controller name. i.e, `ClaimsController` not `ClaimController`

File: app/controllers/api/v1/claim_controller.rb [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    render json: @claim,
+           serializer: ClaimSerializer,
+           root: 'claim'
+  end
+
+  def edit
+  end
+
+  def update
+    @claim.update(claim_params)
```
Comment (vsrboth) [inline app/controllers/api/v1/claim_controller.rb]: prefer `update_attributes` to `update` in order to allow active record to check validation before saving value to table.

File: app/controllers/api/v1/claim_controller.rb [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+    render json: { status: :ok }
+  end
+
+  private
+
+  def claim_params
+    params.require(:claim).permit(:account_id,
+                                  :claimant_id,
+                                  :insurer_id,
+                                  :account,
```
Comment (vsrboth) [inline app/controllers/api/v1/claim_controller.rb]: what is `account` referring to?

File: app/models/claim.rb [hunk 7]
```diff
@@ -0,0 +1,9 @@
+class Claim < ApplicationRecord
+  belongs_to :account
+  belongs_to :insurer
+  belongs_to :claimant, class_name: 'Claimant', foreign_key: :claimant_id, primary_key: :id
+  belongs_to :third_party_claimant, class_name: 'Claimant', foreign_key: :third_party_claimant_id, primary_key: :id
+
+  enum status: [:open, :docs_required, :pending, :in_review, :re_open, :closed]
```
Comment (vsrboth) [inline app/models/claim.rb]: please move `enum` section to above `association`

File: app/models/claim_photo.rb [hunk 8]
```diff
@@ -0,0 +1,11 @@
+class ClaimPhoto < ApplicationRecord
+  after_initialize :defaults, unless: :persisted?
```
Comment (vsrboth) [inline app/models/claim_photo.rb]: please use [Rails observer](https://github.com/rails/rails-observers) for callback.

File: app/models/claim_photo.rb [hunk 9]
```diff
@@ -0,0 +1,11 @@
...
+
+  belongs_to :claim
+  has_attached_file :photo
+  validates_attachment_content_type :photo, content_type: ['image/jpg', 'image/jpeg', 'image/png', 'image/gif']
```
Comment (vsrboth) [inline app/models/claim_photo.rb]: it's better to write this way:

`content_type: %w(image/jpg image/jpeg image/png image/gif)`

File: app/models/claim_photo.rb [hunk 10]
```diff
@@ -0,0 +1,11 @@
...
+
+  def defaults
+    self.photo_taken_date ||= Time.now
```
Comment (vsrboth) [inline app/models/claim_photo.rb]: plz use `Time.current` or `Time.zone.now`

File: app/models/claim.rb (hunk 7 above)
Comment (kduy) [inline app/models/claim.rb]: what do u mean 'above' ?

File: app/controllers/api/v1/claim_photo_controller.rb (hunk 3 above)
Comment (kduy) [inline app/controllers/api/v1/claim_photo_controller.rb]: yeah

File: app/controllers/api/v1/claim_controller.rb (hunk 6 above)
Comment (kduy) [inline app/controllers/api/v1/claim_controller.rb]: i got a prob when testing, then try it out. will remove later.

File: app/models/claim.rb (hunk 7 above)
Comment (vsrboth) [inline app/models/claim.rb]: for  example,

```ruby
enum damage_type: [:water, :storm]

## Associations
belongs_to :account, class_name: 'Account'
```

File: app/controllers/api/v1/claim_controller.rb (hunk 6 above)
Comment (vsrboth) [inline app/controllers/api/v1/claim_controller.rb]: you can't la. You must implement that now 💨

Comment (kduy) [inline app/controllers/api/v1/claim_controller.rb]: i mean , yes

File: app/models/claim.rb (hunk 7 above)
Comment (kduy) [inline app/models/claim.rb]: ok

File: app/models/claim_photo.rb (hunk 10 above)
Comment (kduy) [inline app/models/claim_photo.rb]: why ? 🐤

Comment (vsrboth) [inline app/models/claim_photo.rb]: [Refer to this](http://rails-bestpractices.com/posts/2014/10/22/use-time-zone-now-instead-of-time-now/)

File: app/models/claim_photo.rb (hunk 8 above)
Comment (kduy) [inline app/models/claim_photo.rb]: Rails observer (removed from core in Rails 4.0). should we still use it ?

File: app/controllers/api/v1/claim_controller.rb (hunk 5 above)
Comment (vsrboth) [inline app/controllers/api/v1/claim_controller.rb]: @duyvk can u check this?

File: app/controllers/api/v1/claim_controller.rb (hunk 6 above)
Comment (vsrboth) [inline app/controllers/api/v1/claim_controller.rb]: there is no update on this yet? do u have any problem so far? @duyvk

File: app/controllers/api/v1/claim_controller.rb (hunk 5 above)
Comment (kduy) [inline app/controllers/api/v1/claim_controller.rb]: i already improved

File: app/controllers/api/v1/claim_photo_controller.rb (hunk 3 above)
Comment (vsrboth) [inline app/controllers/api/v1/claim_photo_controller.rb]: there is `index` action yet; anything I can help? @duyvk

File: app/models/claim.rb (hunk 7 above)
Comment (vsrboth) [inline app/models/claim.rb]: there is no update on this yet?

**Comment by kduy** [comment]
new one

**Pull request closed without merging**